File: tests/wasm/test_ma.js
/**
 * WASM binding tests for MA dispatcher.
 * These tests ensure the MA dispatcher correctly routes to various moving average implementations.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('MA dispatcher - SMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "sma", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.sma(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher SMA mismatch");
});

test('MA dispatcher - EMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "ema", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.ema_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher EMA mismatch");
});

test('MA dispatcher - WMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "wma", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.wma_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher WMA mismatch");
});

test('MA dispatcher - ALMA with defaults', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "alma", 9);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.alma_js(close, 9, 0.85, 6.0);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher ALMA mismatch");
});

test('MA dispatcher - DEMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "dema", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.dema_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher DEMA mismatch");
});

test('MA dispatcher - TEMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "tema", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.tema_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher TEMA mismatch");
});

test('MA dispatcher - HMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "hma", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.hma_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher HMA mismatch");
});

test('MA dispatcher - case insensitive', () => {
    const close = new Float64Array(testData.close);
    
    
    const resultUpper = wasm.ma(close, "SMA", 20);
    const resultLower = wasm.ma(close, "sma", 20);
    assertArrayClose(resultUpper, resultLower, 1e-10, "MA dispatcher should be case insensitive");
    
    
    const resultMixed = wasm.ma(close, "EmA", 20);
    const emaResult = wasm.ema_js(close, 20);
    assertArrayClose(resultMixed, emaResult, 1e-10, "MA dispatcher should handle mixed case");
});

test('MA dispatcher - invalid type defaults to SMA', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.ma(close, "invalid_ma_type", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const smaResult = wasm.sma(close, 20);
    assertArrayClose(result, smaResult, 1e-10, "MA dispatcher should default to SMA");
});

test('MA dispatcher - empty data should fail', () => {
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ma(empty, "sma", 20);
    }, /empty|no data/i);
});

test('MA dispatcher - zero period should fail', () => {
    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    assert.throws(() => {
        wasm.ma(data, "sma", 0);
    }, /period/i);
});

test('MA dispatcher - period exceeds length should fail', () => {
    const data = new Float64Array([1.0, 2.0, 3.0]);
    
    assert.throws(() => {
        wasm.ma(data, "sma", 10);
    }, /period|length/i);
});

test('MA dispatcher - all NaN values', () => {
    const allNaN = new Float64Array(100).fill(NaN);
    
    assert.throws(() => {
        wasm.ma(allNaN, "sma", 20);
    }, /NaN|all values/i);
});

test('MA dispatcher - multiple MA types', () => {
    const close = new Float64Array(testData.close.slice(0, 1000)); 
    
    
    const maTypes = [
        "sma", "ema", "dema", "tema", "smma", "zlema", "alma", "cwma",
        "edcf", "fwma", "gaussian", "highpass", "highpass2", "hma",
        "jma", "jsa", "kama", "linreg", "nma", "pwma", "reflex",
        "sinwma", "sqwma", "srwma", "supersmoother", "supersmoother_3_pole",
        "swma", "tilson", "trendflex", "trima", "wilders", "wma"
    ];
    
    for (const maType of maTypes) {
        try {
            const result = wasm.ma(close, maType, 20);
            assert.strictEqual(result.length, close.length, `MA type ${maType} returned wrong length`);
            
            
            let hasValidValues = false;
            for (let i = 100; i < result.length; i++) {
                if (!isNaN(result[i])) {
                    hasValidValues = true;
                    break;
                }
            }
            assert.ok(hasValidValues, `MA type ${maType} returned all NaN`);
        } catch (e) {
            assert.fail(`MA type ${maType} failed: ${e.message}`);
        }
    }
});

test('MA dispatcher - special MA types with defaults', () => {
    const close = new Float64Array(testData.close.slice(0, 1000)); 
    
    
    const hwmaResult = wasm.ma(close, "hwma", 20);
    assert.strictEqual(hwmaResult.length, close.length);
    
    
    const maaqResult = wasm.ma(close, "maaq", 20);
    assert.strictEqual(maaqResult.length, close.length);
    
    
    const mamaResult = wasm.ma(close, "mama", 20);
    assert.strictEqual(mamaResult.length, close.length);
    
    
    const mwdxResult = wasm.ma(close, "mwdx", 20);
    assert.strictEqual(mwdxResult.length, close.length);
    
    
    const ehlersResult = wasm.ma(close, "ehlers_itrend", 50);
    assert.strictEqual(ehlersResult.length, close.length);
});

test('MA dispatcher - KAMA', () => {
    const close = new Float64Array(testData.close);
    
    const result = wasm.ma(close, "kama", 10);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.kama_js(close, 10);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher KAMA mismatch");
});

test('MA dispatcher - ZLEMA', () => {
    const close = new Float64Array(testData.close);
    
    const result = wasm.ma(close, "zlema", 20);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.zlema_js(close, 20);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher ZLEMA mismatch");
});

test('MA dispatcher - Wilders', () => {
    const close = new Float64Array(testData.close);
    
    const result = wasm.ma(close, "wilders", 14);
    assert.strictEqual(result.length, close.length);
    
    
    const directResult = wasm.wilders_js(close, 14);
    assertArrayClose(result, directResult, 1e-10, "MA dispatcher Wilders mismatch");
});

File: tests/python/test_ma.py
"""
Python binding tests for MA dispatcher.
These tests ensure the MA dispatcher correctly routes to various moving average implementations.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close


class TestMaDispatcher:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_ma_sma(self, test_data):
        """Test MA dispatcher with SMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "sma", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.sma(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher SMA mismatch")

    def test_ma_ema(self, test_data):
        """Test MA dispatcher with EMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "ema", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.ema(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher EMA mismatch")

    def test_ma_wma(self, test_data):
        """Test MA dispatcher with WMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "wma", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.wma(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher WMA mismatch")

    def test_ma_alma(self, test_data):
        """Test MA dispatcher with ALMA (uses default offset and sigma)"""
        close = test_data['close']


        result = ta_indicators.ma(close, "alma", 9)
        assert len(result) == len(close)


        direct_result = ta_indicators.alma(close, 9, 0.85, 6.0)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher ALMA mismatch")

    def test_ma_dema(self, test_data):
        """Test MA dispatcher with DEMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "dema", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.dema(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher DEMA mismatch")

    def test_ma_tema(self, test_data):
        """Test MA dispatcher with TEMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "tema", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.tema(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher TEMA mismatch")

    def test_ma_hma(self, test_data):
        """Test MA dispatcher with HMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "hma", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.hma(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher HMA mismatch")

    def test_ma_kama(self, test_data):
        """Test MA dispatcher with KAMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "kama", 10)
        assert len(result) == len(close)


        direct_result = ta_indicators.kama(close, 10)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher KAMA mismatch")

    def test_ma_zlema(self, test_data):
        """Test MA dispatcher with ZLEMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "zlema", 20)
        assert len(result) == len(close)


        direct_result = ta_indicators.zlema(close, 20)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher ZLEMA mismatch")

    def test_ma_wilders(self, test_data):
        """Test MA dispatcher with Wilders"""
        close = test_data['close']


        result = ta_indicators.ma(close, "wilders", 14)
        assert len(result) == len(close)


        direct_result = ta_indicators.wilders(close, 14)
        assert_close(result, direct_result, rtol=1e-10, msg="MA dispatcher Wilders mismatch")

    def test_ma_invalid_type(self, test_data):
        """Test MA dispatcher with invalid MA type defaults to SMA"""
        close = test_data['close']


        result = ta_indicators.ma(close, "invalid_ma_type", 20)
        assert len(result) == len(close)


        sma_result = ta_indicators.sma(close, 20)
        assert_close(result, sma_result, rtol=1e-10, msg="MA dispatcher should default to SMA")

    def test_ma_case_insensitive(self, test_data):
        """Test MA dispatcher is case insensitive"""
        close = test_data['close']


        result_upper = ta_indicators.ma(close, "SMA", 20)
        result_lower = ta_indicators.ma(close, "sma", 20)
        assert_close(result_upper, result_lower, rtol=1e-10, msg="MA dispatcher should be case insensitive")


        result_mixed = ta_indicators.ma(close, "EmA", 20)
        ema_result = ta_indicators.ema(close, 20)
        assert_close(result_mixed, ema_result, rtol=1e-10, msg="MA dispatcher should handle mixed case")

    def test_ma_empty_input(self):
        """Test MA dispatcher fails with empty input"""
        empty = np.array([])

        with pytest.raises(ValueError):
            ta_indicators.ma(empty, "sma", 20)

    def test_ma_zero_period(self):
        """Test MA dispatcher fails with zero period"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])

        with pytest.raises(ValueError):
            ta_indicators.ma(data, "sma", 0)

    def test_ma_period_exceeds_length(self):
        """Test MA dispatcher fails when period exceeds data length"""
        data = np.array([1.0, 2.0, 3.0])

        with pytest.raises(ValueError):
            ta_indicators.ma(data, "sma", 10)

    def test_ma_all_nan_input(self):
        """Test MA dispatcher with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError):
            ta_indicators.ma(all_nan, "sma", 20)

    def test_ma_all_supported_types(self, test_data):
        """Test MA dispatcher with all supported MA types"""
        close = test_data['close'][:1000]


        ma_types = [
            "sma", "ema", "dema", "tema", "smma", "zlema", "alma", "cwma",
            "edcf", "fwma", "gaussian", "highpass", "highpass2", "hma",
            "jma", "jsa", "kama", "linreg", "nma", "pwma", "reflex",
            "sinwma", "sqwma", "srwma", "supersmoother", "supersmoother_3_pole",
            "swma", "tilson", "trendflex", "trima", "wilders", "wma"
        ]

        for ma_type in ma_types:
            try:
                result = ta_indicators.ma(close, ma_type, 20)
                assert len(result) == len(close), f"MA type {ma_type} returned wrong length"

                assert not np.all(np.isnan(result[100:])), f"MA type {ma_type} returned all NaN"
            except Exception as e:
                pytest.fail(f"MA type {ma_type} failed: {str(e)}")

    def test_ma_special_types_with_defaults(self, test_data):
        """Test MA types that have special default handling"""
        close = test_data['close'][:1000]


        result = ta_indicators.ma(close, "hwma", 20)
        assert len(result) == len(close)


        result = ta_indicators.ma(close, "maaq", 20)
        assert len(result) == len(close)


        result = ta_indicators.ma(close, "mama", 20)
        assert len(result) == len(close)


        result = ta_indicators.ma(close, "mwdx", 20)
        assert len(result) == len(close)


        result = ta_indicators.ma(close, "ehlers_itrend", 50)
        assert len(result) == len(close)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])